### PR TITLE
[Merged by Bors] - Utility methods for Val

### LINKED ISSUE #6080: Add utility methods to `Val`
## What problem does this solve or what need does it fill?

The [`Val`](https://docs.rs/bevy/latest/bevy/prelude/enum.Val.html) type is an expressive way to represent UI scales across unit types.

However, working with these values can be quite frustrating; basic utility methods are missing.

## What solution would you like?

1. Add `try_add` and `try_sub` methods to `Val`, which only return Ok if the enum variants are identical.
2. Remove the `Add` and `AddAssign` impls for `Val`.
3. Add `evaluate(size: f32)` method that converts from `Val::Percent` to `Val::Px`.
4. Add `try_add_with_size` and `try_sub_with_size` methods to `Val`, which evaluate `Val::Percent` values into `Val::Px` values before adding.

## What alternative(s) have you considered?

We could implement the `Add` and `Sub` traits (and already implement the `Add` trait).

However, this results in some fairly tricky implicit behavior. For example, what happens when we add `Val::Undefined` to `Val::Auto`? This implicit introduction of `NaN`-like behavior can be very frustrating to debug.

### PR DESCRIPTION
**Pull request by Pietrek14**
# Objective

Adds a better interface for performing mathematical operations with UI unit `Val`. Fixes #6080.

## Solution

- Added `try_add` and `try_sub` methods to Val.
- Removed the `Add` and `AddAssign` impls for `Val` that introduced unintuitive and bug-prone behaviour.
- As a consequence of the prior,  ~~changed the `Add` and `Sub` impls for the `Size` struct to take a `(Val, Val)` instead of `Vec2`~~ deleted the `Add` and `Sub` impls for the `Size` struct
- Added a `From<(Val, Val)>` impl for the `Size` struct
- Added `evaluate(size: f32)` method that converts from `Val::Percent` to `Val::Px`.
- Added `try_add_with_size` and `try_sub_with_size` methods to `Val`, which evaluate `Val::Percent` values into `Val::Px` values before adding.

---

## Migration Guide

Instead of using the + and - operators, perform calculations on `Val`s using the new `try_add` and `try_sub` methods. Multiplication and division remained unchanged. Also, when adding or subtracting from `Size`, ~~use a `Val` tuple instead of `Vec2`~~ perform the addition on `width` and `height` separately.

### CODE REVIEW

File: crates/bevy_ui/src/ui_node.rs [hunk 1]
```diff
@@ -34,91 +35,134 @@ pub enum Val {
...
+    pub fn evaluate(&self, size: f32) -> Result<Val, ValArithmeticError> {
         match self {
-            Val::Undefined => Val::Undefined,
-            Val::Auto => Val::Auto,
-            Val::Px(value) => Val::Px(value / rhs),
-            Val::Percent(value) => Val::Percent(value / rhs),
+            Val::Percent(value) => Ok(Val::Px(size * value / 100.0)),
+            Val::Px(_) => Ok(*self),
+            _ => Err(ValArithmeticError::NonEvaluateable),
         }
```
Comment (tguichaoua) [inline crates/bevy_ui/src/ui_node.rs:147]: I don't think this method should be fallible.
Evaluating `Auto` returns `Auto` and `Undefined` returns `Undefined` looks more intuitive.

File: crates/bevy_ui/src/ui_node.rs [hunk 2]
```diff
@@ -34,91 +35,134 @@ pub enum Val {
...
+    /// When adding non-numeric [`Val`]s, it returns the value unchanged.
+    pub fn try_add(&self, rhs: Val) -> Result<Val, ValArithmeticError> {
+        match (self, rhs) {
+            (Val::Undefined, Val::Undefined) => {
+                warn!("Adding to Val::Undefined is a redundant operation");
+                Ok(*self)
+            }
+            (Val::Auto, Val::Auto) => {
+                warn!("Adding to Val::Auto is a redundant operation");
+                Ok(*self)
```
Comment (tguichaoua) [inline crates/bevy_ui/src/ui_node.rs:100]: Are warnings necessary ?
It may generate log pollution.

File: crates/bevy_ui/src/ui_node.rs [hunk 3]
```diff
@@ -34,91 +35,134 @@ pub enum Val {
...
+    /// When adding non-numeric [`Val`]s, it returns the value unchanged.
+    pub fn try_sub(&self, rhs: Val) -> Result<Val, ValArithmeticError> {
+        match (self, rhs) {
+            (Val::Undefined, Val::Undefined) => {
+                warn!("Subtracting from Val::Undefined is a redundant operation");
+                Ok(*self)
+            }
+            (Val::Auto, Val::Auto) => {
+                warn!("Subtracting from Val::Auto is a redundant operation");
+                Ok(*self)
```
Comment (tguichaoua) [inline crates/bevy_ui/src/ui_node.rs:124]: Are warnings necessary ?
It may generate log pollution.

File: crates/bevy_ui/src/ui_node.rs [hunk 4]
```diff
@@ -34,91 +35,134 @@ pub enum Val {
...
         }
     }
-}
 
-impl MulAssign<f32> for Val {
-    fn mul_assign(&mut self, rhs: f32) {
-        match self {
-            Val::Undefined | Val::Auto => {}
-            Val::Px(value) | Val::Percent(value) => *value *= rhs,
+    pub fn try_add_to_self(&mut self, rhs: Val) -> Result<(), ValArithmeticError> {
```
Comment (tguichaoua) [inline crates/bevy_ui/src/ui_node.rs:108]: Can named `try_add_assign` to mimic the standard api.

File: crates/bevy_ui/src/ui_node.rs [hunk 5]
```diff
@@ -34,91 +35,134 @@ pub enum Val {
...
+            (Val::Px(value), Val::Px(rhs_value)) => Ok(Val::Px(value - rhs_value)),
+            (Val::Percent(value), Val::Percent(rhs_value)) => Ok(Val::Percent(value - rhs_value)),
+            _ => Err(ValArithmeticError::NonIdenticalVariants),
         }
     }
-}
 
-impl Div<f32> for Val {
-    type Output = Val;
+    pub fn try_sub_from_self(&mut self, rhs: Val) -> Result<(), ValArithmeticError> {
```
Comment (tguichaoua) [inline crates/bevy_ui/src/ui_node.rs:132]: Can named `try_sub_assign` to mimic the standard api.

File: crates/bevy_ui/src/ui_node.rs [hunk 6]
```diff
@@ -34,91 +35,134 @@ pub enum Val {
...
 
-impl DivAssign<f32> for Val {
-    fn div_assign(&mut self, rhs: f32) {
-        match self {
-            Val::Undefined | Val::Auto => {}
-            Val::Px(value) | Val::Percent(value) => *value /= rhs,
-        }
+    /// Similar to [`Val::try_add`], but performs [`Val::evaluate`] on both values before adding.
+    /// Both values have to be evaluatable (numeric).
+    pub fn try_add_with_size(&self, rhs: Val, size: f32) -> Result<Val, ValArithmeticError> {
```
Comment (tguichaoua) [inline crates/bevy_ui/src/ui_node.rs:152]: To be consistent you should add `try_add_assign_with_size` and `try_sub_assign_with_size`.

File: crates/bevy_ui/src/geometry.rs [hunk 7]
```diff
@@ -356,39 +355,39 @@ impl Size {
     };
 }
 
-impl Add<Vec2> for Size {
+impl Add<(Val, Val)> for Size {
     type Output = Size;
 
-    fn add(self, rhs: Vec2) -> Self::Output {
+    fn add(self, rhs: (Val, Val)) -> Self::Output {
```
Comment (tguichaoua) [inline crates/bevy_ui/src/geometry.rs:361]: I think you should add doc with a "panic" section to mention cases this method may panic.

Same for the other operator methods.

**Review by tguichaoua — CHANGES_REQUESTED** [review]

File: crates/bevy_ui/src/ui_node.rs (hunk 1 above)
Comment (Pietrek14) [reply crates/bevy_ui/src/ui_node.rs:147]: The point of this function is to return the actual `Px` value that the `Val` is going to be evaluated to, so returning `Auto` or `Undefined` wouldn't make sense imo.

Comment (Pietrek14) [reply crates/bevy_ui/src/ui_node.rs:147]: Yeah, that makes sense.

**Comment by alice-i-cecile** [comment]
@Pietrek14 feel free to resolve comments as you address them; it makes it much easier for other reviewers to see what still needs to be done :)

**Comment by Pietrek14** [comment]
I think I covered it all. Sorry it took so long, I was busy with Ludum Dare.

File: crates/bevy_ui/src/geometry.rs [hunk 8]
```diff
@@ -356,39 +355,55 @@ impl Size {
     };
 }
 
-impl Add<Vec2> for Size {
+impl Add<(Val, Val)> for Size {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/geometry.rs:358]: Taking a look, I don't like these impls (or the previous ones for Vec2).

Addition and subtraction are common, hard-to-grep operations: they should not implicitly panic.

Instead, I would just remove these, and add a `From<(Val, Val)> for Size` impl if it doesn't exist already.

File: crates/bevy_ui/src/ui_node.rs [hunk 9]
```diff
@@ -34,91 +34,142 @@ pub enum Val {
...
             Val::Undefined | Val::Auto => {}
-            Val::Px(value) | Val::Percent(value) => *value -= rhs,
+            Val::Px(value) | Val::Percent(value) => *value /= rhs,
         }
     }
 }
 
-impl Mul<f32> for Val {
-    type Output = Val;
+#[derive(Debug, Eq, PartialEq)]
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/ui_node.rs:81]: Clone and Copy too :)

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]

**Comment by Pietrek14** [comment]
Should I adjust the migration guide in the PR description, since I deleted the addition and subtraction of `Size`?

**Comment by alice-i-cecile** [comment]
Yes please :)

File: crates/bevy_ui/src/ui_node.rs [hunk 10]
```diff
@@ -34,91 +35,144 @@ pub enum Val {
...
         }
     }
-}
 
-impl MulAssign<f32> for Val {
-    fn mul_assign(&mut self, rhs: f32) {
-        match self {
-            Val::Undefined | Val::Auto => {}
-            Val::Px(value) | Val::Percent(value) => *value *= rhs,
+    pub fn try_add_assign(&mut self, rhs: Val) -> Result<(), ValArithmeticError> {
```
Comment (tguichaoua) [inline crates/bevy_ui/src/ui_node.rs:103]: Maybe add a doc to at least say it's a assign version of `try_add`.

File: crates/bevy_ui/src/ui_node.rs [hunk 11]
```diff
@@ -34,91 +35,144 @@ pub enum Val {
...
+            (Val::Px(value), Val::Px(rhs_value)) => Ok(Val::Px(value - rhs_value)),
+            (Val::Percent(value), Val::Percent(rhs_value)) => Ok(Val::Percent(value - rhs_value)),
+            _ => Err(ValArithmeticError::NonIdenticalVariants),
         }
     }
-}
 
-impl Div<f32> for Val {
-    type Output = Val;
+    pub fn try_sub_assign(&mut self, rhs: Val) -> Result<(), ValArithmeticError> {
```
Comment (tguichaoua) [inline crates/bevy_ui/src/ui_node.rs:120]: ditto

**Review by tguichaoua — CHANGES_REQUESTED** [review]
There is two methods without doc.

Other than that LGTM.

**Review by tguichaoua — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]
Great work, I'm happy with how this turned out. Love the error enum. We can add more impls and methods later once we see how this batch plays out.

bors r+

**Comment by Pietrek14** [comment]
It's really not that important, but do PRs merged by bors count for Hacktoberfest?

**Pull request closed without merging**